substrate/client/consensus/babe/src/lib.rs

```diff
@@ -1388,10 +1230,125 @@ where
 
 		Ok(ImportResult::Imported(aux))
 	}
+
+	async fn check_inherents(
+		&self,
+		block: Block,
+		at_hash: Block::Hash,
+		inherent_data: InherentData,
+		create_inherent_data_providers: CIDP::InherentDataProviders,
+	) -> Result<(), ConsensusError> {
+		let inherent_res = self
+			.client
+			.runtime_api()
+			.check_inherents(at_hash, block, inherent_data)
+			.map_err(|e| ConsensusError::Other(Box::new(e)))?;
+
+		if !inherent_res.ok() {
+			for (i, e) in inherent_res.into_errors() {
+				match create_inherent_data_providers.try_handle_error(&i, &e).await {
+					Some(res) => res.map_err(|e| ConsensusError::InvalidInherents(e))?,
+					None => return Err(ConsensusError::InvalidInherentsUnhandled(i)),
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn check_and_report_equivocation(
```

Comment:
While this is fine to have it here right now. This should be some background process to which the verify function is sending the headers to check. The point being that you already want to report equivocations when you see the header not want to wait for receiving the entire body (which may never appears). 

(This can be some followup)